### DISCUSSION TITLE
How does folder with LUA scripts work?

### DISCUSSION THREAD

**Discussion by raffaelemancuso** [L3MON4D3/LuaSnip#653] [Q&A]
I'm reading the suggested [A LuaSnip guide for LaTeX workflows](https://www.ejmastnak.com/tutorials/vim-latex/luasnip.html).

I have created the file `~/.config/nvim/snippets/my-snippets/python.lua` and I have put this inside of it:
```
require("luasnip").snippet(
	{ -- Table 1: snippet parameters
          trig="hi",
          dscr="An autotriggering snippet that expands 'hi' into 'Hello, world!'",
          regTrig=false,
          priority=100,
          snippetType="autosnippet"
	},
	{ -- Table 2: snippet nodes (don't worry about this for now---we'll cover nodes shortly)
          t("Hello, world!"), -- A single text node
	}
	-- Table 3, the advanced snippet options, is left blank.
)
```

In my `init.lua` I have:
```
require("luasnip.loaders.from_lua").load({paths = HOME .. "/.config/nvim/snippets/my-snippets"})
```

When I type "hi", nothing happens.

If I wrap the content of `python.lua` inside `add_snippets`:
```
local ls = require("luasnip")
ls.add_snippets("python", {
	require("luasnip").snippet(
	  { -- Table 1: snippet parameters
		trig="hi",
		dscr="An autotriggering snippet that expands 'hi' into 'Hello, world!'",
		regTrig=false,
		priority=100,
		snippetType="autosnippet"
	  },
	  { -- Table 2: snippet nodes (don't worry about this for now---we'll cover nodes shortly)
		t("Hello, world!"), -- A single text node
	  }
	  -- Table 3, the advanced snippet options, is left blank.
	),
})
```

Then it works correctly.

I don't understand why the guide didn't tell me to do so.

I also don't understand if "The file’s base name determines which Vim filetype the snippets apply to. For example, snippets inside the file tex.lua would apply to files with filetype=tex".

If that is the case, do I need to specify `python` as the first argument of `add_snippets`? isn't the filename sufficient to identify which filetype the snippet applies to?

**Comment by L3MON4D3** [comment]
Ah, that article omits one detail: you'll have to also return the defined snippets in that file
```lua
-- python.lua
return { require("luasnip").snippet(...) }
```
It also works with `add_snippets` since that's the function that actually adds snippets (and that file is run like any other lua-file, so calling that in there works, but defeats the purpose of the loader)

**Reply by raffaelemancuso** [reply]
ok. Must the file be named `python.lua`? If so, what's the purpose of the "python" in `ls.add_snippets("python"`? Isn't that an useless repetition?

**Reply by L3MON4D3** [reply]
Yup, the filename should be `python.lua`.
You don't have to call `ls.add_snippets(filetype, {<snippets>})` in that file, you can just `return {<snippets>}`

**Reply by raffaelemancuso** [reply]
May I say the docs for newbies need a rewrite?

We first make it too complex by listing all possible snippets formats and how they are loaded. 

Then we rely on external sources which are wrong ("that article omits one detail", as you said) and may become outdated.

I would write a "Getting started" page in the wiki, with a step-by-step instructions to get a very simple first script up and running.

Then we may link to external sources with a warning about their errors.

**Reply by L3MON4D3** [reply]
> May I say the docs for newbies need a rewrite?

You may :P
TBH getting new users up to speed is pretty hard, luasnip is huuuge, and pretty different from other snippet-plugins.
IMO the videos by TJ are a good first introduction, even though some api has changed since they were made, the fundamentals are well explained, and still (and probably forever) valid.

> We first make it too complex by listing all possible snippets formats and how they are loaded.

What part of the doc do you mean by that? The `Getting started`-part in the README?
I think it's important to show the other formats, since the lua-snippets are less comfortable to write, and if their capabilities are not necessary, writing snipmate-snippets first is the better choice.

> Then we rely on external sources which are wrong ("that article omits one detail", as you said) and may become outdated.

Mhmm true becoming outdated is a problem, but so far that isn't much of an issue (the underlying concepts don't change, only (very seldomly, I might add) the interface, and that is well documented in `DOC.md`)
But, yeah, that could be better-explained in the article.
@ejmastnak WDYT about showing a minimal lua-loaded file at some (early) point in your guide? 

> I would write a "Getting started" page in the wiki, with a step-by-step instructions to get a very simple first script up and running.

The official doc actually contains a complete example for [getting the lua-loader to work](https://github.com/L3MON4D3/LuaSnip/blob/master/DOC.md#lua) (actually, for all loaders), which is linked in the README.
There's also `Examples/snippets.lua`, which is mentioned in the readme, and should always work.

I feel like the main issue is that there's just this big wall of text, of resources/information, which somewhat encourages just skipping through it.

**Reply by ejmastnak** [reply]
@raffaem thank you very much for pointing out the missing return statement. That's a serious oversight on my part and I will fix it in the next few hours.

>  @ejmastnak WDYT about showing a minimal lua-loaded file at some (early) point in your guide?

Yes, good idea! Should be done by tonight.

**Reply by ejmastnak** [reply]
Okay, here's a first iteration: https://www.ejmastnak.com/tutorials/vim-latex/luasnip.html#tldr-hello-world-example

@raffaem, since you inspired this perhaps you could test it and give your opinion? E.g. does the hello-world example work, did I miss some obvious detail again, is the example concise enough, etc

**Reply by L3MON4D3** [reply]
As far as I'm concerned, that's pretty good :+1:
Nits/minor suggestions:
* in 2 the order is `init.vim`, `init.lua`, reversed in 4 (not sure if you care about stuff like this xD)
* 5: maybe "open a new neovim-instance" instead of opening a file since the file isn't necessary for this to work
* I feel like 6 and 7 (maybe even 5,6,7) could be collapsed into one point, "enter insert, type hi, expand -> hello world"

Oh, and really nice that you could get to that so quickly :)

**Reply by ejmastnak** [reply]
>  Nits/minor suggestions:

Thanks! I think all of these are fair points, just updated the guide.

> Oh, and really nice that you could get to that so quickly :)

Haha otherwise I would probably forget and never get around to it :P

**Reply by raffaelemancuso** [reply]
@ejmastnak it is fair now. Thank you so much for your work.

Altough I still don't understand the format of the return statement.

Does it return a **list** (what LUA calls table) of snippets?

So I can return more than one snippet?

Maybe we should specify this :)

**Reply by ejmastnak** [reply]
> @ejmastnak it is fair now. Thank you so much for your work.

Great, glad to hear.

> Does it return a list (what LUA calls table) of snippets?
> So I can return more than one snippet?

Yes, exactly, to return more than one snippet you return a Lua table/list of snippets (using only one `return` per file). For example:

   ```lua
   -- Place this in ${HOME}/.config/nvim/LuaSnip/all.lua
   return {
     -- A snippet that expands the trigger "hi" into the string "Hello, world!".
     require("luasnip").snippet(
       { trig = "hi" },
       { t("Hello, world!") }
     ),

     -- To return multiple snippets, use one `return` statement per snippet file
     -- and return a table of Lua snippets.
     require("luasnip").snippet(
       { trig = "foo" },
       { t("Another snippet.") }
     ),
   }
   ```

I also just updated the blog post with the above two-snippet example.
